### PR TITLE
refactor bytes

### PR DESCRIPTION
**Pull request by jgirardet**
It's the start to refactor bytes and  bytearray like #765 .

- only bytes for now (no bytearray) with new style method
- What to do with byteinner ? change the name ? remove a differet file and put everything in bytes.rs
- tests added
- I struggled with the iterator, I don't know if it can be send to byteinner.rs or not.
- the `__new__` method has been rewritten to deal with more cases, but it's quite horrible to read. interested in ideas to refactor this one. there are many possibilities with various error message so ...
- `b"éùà"` passes now but it shouldn't. I'm not sur where to fix it.

### CODE REVIEW

File: vm/src/obj/objbyteinner.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                            enc.class().name
+                        )));
+                    }
+                } else {
+                    return Err(vm.new_type_error("encoding without a string argument".to_string()));
+                }
+            } else {
+                return Err(vm.new_type_error("encoding without a string argument".to_string()));
+            }
+        // On ly one argument
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs]: typo: `Only`

File: vm/src/obj/objbyteinner.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    pub fn add(&self, other: &PyByteInner, vm: &VirtualMachine) -> PyResult {
+        let elements: Vec<u8> = self
+            .elements
+            .iter()
+            .chain(other.elements.iter())
+            .cloned()
+            .collect();
+        Ok(vm.ctx.new_bytes(elements))
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs]: This should probably return a new `PyByteInner` and not `bytes`.

Comment (jgirardet) [reply vm/src/obj/objbyteinner.rs]: I finally did it with Vec[u8] instead , easier to call wm.ctx.new_bytes()

Comment (jgirardet) [reply vm/src/obj/objbyteinner.rs]: But thinking to this, maybe it's more consistent to always return a PyByteInner when it's possible. Anyone has an opinion on it ?

**Comment by coolreader18** [comment]
The tests where you're `except`ing and then `assert`ing true could be changed to use `assertRaises`. Other than that, LGTM

File: vm/src/obj/objbyteinner.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+    pub fn contains_int(&self, int: &PyInt, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        if let Some(int) = int.as_bigint().to_u8() {
+            if self.elements.contains(&int) {
+                Ok(vm.new_bool(true))
+            } else {
+                Ok(vm.new_bool(false))
+            }
+        } else {
+            Err(vm.new_value_error("byte mu st be in range(0, 256)".to_string()))
```
Comment (adrian17) [inline vm/src/obj/objbyteinner.rs]: typo

File: vm/src/obj/objbyteinner.rs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+            .chunks(2)
+            .map(|x| x.to_vec().iter().collect::<String>())
+            .map(|x| u8::from_str_radix(&x, 16))
+            .map(|x| x.unwrap())
+            .collect::<Vec<u8>>())
+    }
+}
+
+// TODO
+// fix b"é" not allowed should be bytes("é", "utf8")
```
Comment (adrian17) [inline vm/src/obj/objbyteinner.rs]: it's not allowed anymore, right? Is this TODO still relevant?

File: vm/src/obj/objbyteinner.rs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        // First handle bytes(string, encoding[, errors])
+        if let OptionalArg::Present(enc) = enc_option {
+            if let OptionalArg::Present(eval) = val_option {
+                if let Ok(input) = eval.downcast::<PyString>() {
+                    if let Ok(encoding) = enc.clone().downcast::<PyString>() {
+                        if encoding.value.to_lowercase() == "utf8".to_string()
+                            || encoding.value.to_lowercase() == "utf-8".to_string()
+                        // TODO: different encoding
+                        {
+                            return Ok(PyByteInner {
```
Comment (adrian17) [inline vm/src/obj/objbyteinner.rs]: Could you run clippy on this? These seem like redundant `return`s.

File: vm/src/obj/objbyteinner.rs (hunk 4 above)
Comment (jgirardet) [reply vm/src/obj/objbyteinner.rs]: no it's ok now

File: vm/src/obj/objbyteinner.rs (hunk 5 above)
Comment (jgirardet) [reply vm/src/obj/objbyteinner.rs]: ok

**Comment by jgirardet** [comment]
ok changes done

**Pull request merged**